crates/ra_mbe/src/subtree_source.rs

```diff
@@ -129,7 +129,8 @@ fn convert_delim(d: Option<tt::DelimiterKind>, closing: bool) -> TtToken {
 }
 
 fn convert_literal(l: &tt::Literal) -> TtToken {
-    let kind = lex_single_valid_syntax_kind(&l.text)
+    let kind = lex_single_syntax_kind(&l.text)
```

Comment:
What is the different between `lex_single_syntax_kind` and `lex_single_valid_syntax_kind` ?

Reply:
There is a documentation:
https://github.com/rust-analyzer/rust-analyzer/blob/20252efb32bfdfe7392934a95a6c6d6b583d10e7/crates/ra_syntax/src/parsing/lexer.rs#L59-L76

Reply:
So if I understand correctly, `lex_single_syntax_kind` allows syntax error but still return the SyntaxKind back such that we treat`"\");\n}\n"` is a valid literal here, right ? I don't know whether if other parts of `mbe` could handle this "invalid" valid literal ...

Reply:
They should, before we introduced collecting lexer errors all of them were ignored everywhere, so ra_mbe handled malformed tokens correctly.

Reply:
Sounds reasonable.

it would be nice to add a test in :

https://github.com/rust-analyzer/rust-analyzer/blob/20252efb32bfdfe7392934a95a6c6d6b583d10e7/crates/ra_mbe/src/tests.rs#L1524

Something like:

```
#[test]
fn test_expand_bad_literal() {
    // FIXME: the second rule of the macro should be removed and an error about
    // `$( $c )+` raised
    parse_macro(
        r#"
        macro_rules! foo { ($i:literal) => {}; }
    "#,
    )
    .assert_expand_items("foo!(&k");", "");
}

Reply:
Sounds reasonable !